keps/sig-storage/1472/README.md

```diff
@@ -0,0 +1,1349 @@
+with a CSI controller service and capacity reporting gets enabled with
+`--enable-capacity=central`. In this mode, CSI drivers must report
+topology information in `NodeGetInfoResponse.accessible_topology` that
+matches the storage pool(s) that it has access to, with granularity
+that matches the most restrictive pool.
+
+For example, if the driver runs in a node with zone/region/rack
+topology and has access to one per-zone pool and one per-zone/region
+pool, then the driver should report the zone/region of the second pool
+as its topology.
+
+This is a slight deviation from the CSI spec, which defines that
+"`accessible_topology` specifies where the *node* is accessible
+from". For node-local storage there is no difference, but for
+network-attached storage there might be one.
+
+Assuming that a CSI driver meets the above requirement and enables
+this mode, external-provisioner then can identify pools as follows:
+- iterate over all `CSINode` objects and search for
+  `CSINodeDriver` information for the CSI driver,
+- compute the union of the topology segments from these
+  `CSINodeDriver` entries.
+
+For each entry in that union, one `CSIStorageCapacity` object is
+created for each storage class where the driver reports a non-zero
+capacity. The node selector uses the topology key/value pairs as node
+labels. That works because kubelet automatically labels nodes based on
+the CSI drivers that run on that node.
+
+#### With central controller, generic solution
+
+For all other cases, a new CSI call to enumerate storage topology
+segments will be needed. That call then might also include other
+information for each segment.
+
+#### Determining parameters
+
+After determining the topology as described above,
+external-provisioner needs to figure out with which volume parameters
+it needs to call `GetCapacity`. It does that by iterating over all
+storage classes that reference the driver.
+
+If the current combination of topology segment and storage class
+parameters do not make sense, then a driver must return "zero
+capacity" or an error, in which case external-provisioner will skip
+this combination. This covers the case where some storage class
+parameter limits the topology segment of volumes using that class,
+because information will then only be recorded for the segments that
+are supported by the class.
+
+#### CSIStorageCapacity lifecycle
+
+external-provisioner needs permission to create, update and delete
+`CSIStorageCapacity` objects. Before creating a new object, it must
+check whether one already exists with the relevant attributes (driver
+name + nodes + storage class) and then update that one
+instead. Obsolete objects need to be removed.
+
+To ensure that `CSIStorageCapacity` objects get removed when the driver
+deployment gets removed before it has a chance to clean up, each
+`CSIStorageCapacity` object needs an [owner
+reference](https://godoc.org/k8s.io/apimachinery/pkg/apis/meta/v1#OwnerReference).
```

Comment:
Yes, it is possible to identify the pod and thus its owners from inside the pod itself. We've implemented that for PMEM-CSI to find the namespace in which our operator runs, so I know this can work.